File: zig/test/renderToConsole_test.zig
const std = @import("std");

const parse = @import("root.zig").parse.execute;
const core = @import("rulesets/core.zig").core;
const gfm = @import("rulesets/gfm.zig").gfm;
const renderToConsole = @import("root.zig").render.renderToConsole;

test "renderConsole paragraph" {
    const input = "Hello, world!";
    const root = try parse(std.testing.allocator, input, core.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, core.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "Hello, world!") != null);
}

test "renderConsole heading" {
    const input = "# Heading 1\n## Heading 2";
    const root = try parse(std.testing.allocator, input, core.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, core.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "# Heading 1") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "## Heading 2") != null);
}

test "renderConsole bulleted list" {
    const input = "- Item 1\n- Item 2";
    const root = try parse(std.testing.allocator, input, core.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, core.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "• Item 1") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "• Item 2") != null);
}

test "renderConsole ordered list" {
    const input = "1. First\n2. Second";
    const root = try parse(std.testing.allocator, input, core.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, core.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "1. First") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "2. Second") != null);
}

test "renderConsole code block" {
    const input = "```\ncode\n```";
    const root = try parse(std.testing.allocator, input, core.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, core.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "┌─") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "│") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "└─") != null);
}

test "renderConsole inline code" {
    const input = "`code`";
    const root = try parse(std.testing.allocator, input, core.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, core.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "`code`") != null);
}

test "renderConsole block quote" {
    const input = "> Quote text";
    const root = try parse(std.testing.allocator, input, core.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, core.ruleSet);
    const stripped = stripAnsi(std.testing.allocator, output);
    defer std.testing.allocator.free(stripped);
    try std.testing.expect(std.mem.indexOf(u8, stripped, "┃ Quote text") != null);
}

fn stripAnsi(allocator: std.mem.Allocator, input: []const u8) []const u8 {
    var result = std.ArrayList(u8).initCapacity(allocator, input.len) catch unreachable;
    defer result.deinit(allocator);
    var i: usize = 0;
    while (i < input.len) {
        if (input[i] == 0x1B and i + 1 < input.len and input[i + 1] == '[') {
            var j = i + 2;
            while (j < input.len) {
                if (input[j] == 'm') {
                    i = j + 1;
                    break;
                }
                if (!std.ascii.isDigit(input[j]) and input[j] != ';') {
                    break;
                }
                j += 1;
            }
            continue;
        }
        result.append(allocator, input[i]) catch unreachable;
        i += 1;
    }
    return result.toOwnedSlice(allocator) catch unreachable;
}

test "renderConsole thematic break" {
    const input = "---";
    const root = try parse(std.testing.allocator, input, core.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, core.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "─") != null);
}

test "renderConsole task list" {
    const input = "- [x] Done\n- [ ] Todo";
    const root = try parse(std.testing.allocator, input, gfm.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, gfm.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "[✓]") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "[ ]") != null);
}

test "renderConsole table" {
    const input = "| A | B |\n|---|---|\n| 1 | 2 |";
    const root = try parse(std.testing.allocator, input, gfm.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, gfm.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "┌") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "┬") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "┐") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "┼") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "│") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "├") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "┤") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "└") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "┴") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "┘") != null);
}

test "renderConsole strong text" {
    const input = "**bold**";
    const root = try parse(std.testing.allocator, input, core.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, core.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "bold") != null);
}

test "renderConsole emphasis text" {
    const input = "*italic*";
    const root = try parse(std.testing.allocator, input, core.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, core.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "italic") != null);
}

test "renderConsole link" {
    const input = "[text](url)";
    const root = try parse(std.testing.allocator, input, core.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, core.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "text") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "url") != null);
}

test "renderConsole image" {
    const input = "![alt](url)";
    const root = try parse(std.testing.allocator, input, core.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, core.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "Image") != null);
}

test "renderConsole strikethrough" {
    const input = "~~deleted~~";
    const root = try parse(std.testing.allocator, input, gfm.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, gfm.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "~~deleted~~") != null);
}

test "renderConsole alert" {
    const input = "> [!NOTE]\n> Note content";
    const root = try parse(std.testing.allocator, input, gfm.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, gfm.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "📝") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "Note:") != null);
}

test "renderConsole nested list" {
    const input = "- Level 1\n  - Level 2\n    - Level 3";
    const root = try parse(std.testing.allocator, input, core.ruleSet, false);
    const output = try renderToConsole(std.testing.allocator, root, core.ruleSet);
    try std.testing.expect(std.mem.indexOf(u8, output, "• Level 1") != null);
    try std.testing.expect(std.mem.indexOf(u8, output, "◦ Level 2") != null);
}

File: web/test/renderToConsole.test.ts
import { expect, test } from "vitest";
import parse from "../src/parse";
import renderToConsole from "../src/renderToConsole";
import core from "../src/rulesets/core";
import gfm from "../src/rulesets/gfm";

test("renders paragraph to console", () => {
	const input = "Hello, world!";
	const root = parse(input, core, false);
	const output = renderToConsole(root);
	expect(output).toBe("Hello, world!");
});

test("renders heading to console with color", () => {
	const input = "# Heading 1\n## Heading 2";
	const root = parse(input, core, false);
	const output = renderToConsole(root);
	expect(output).toContain("# Heading 1");
	expect(output).toContain("## Heading 2");
});

test("renders bulleted list with Unicode bullets", () => {
	const input = "- Item 1\n- Item 2";
	const root = parse(input, core, false);
	const output = renderToConsole(root);
	expect(output).toContain("• Item 1");
	expect(output).toContain("• Item 2");
});

test("renders ordered list", () => {
	const input = "1. First\n2. Second";
	const root = parse(input, core, false);
	const output = renderToConsole(root);
	expect(output).toContain("1. First");
	expect(output).toContain("2. Second");
});

test("renders code block with box drawing", () => {
	const input = "```\ncode\n```";
	const root = parse(input, core, false);
	const output = renderToConsole(root);
	expect(output).toContain("┌─");
	expect(output).toContain("│");
	expect(output).toContain("└─");
});

test("renders inline code", () => {
	const input = "`code`";
	const root = parse(input, core, false);
	const output = renderToConsole(root);
	expect(output).toContain("`code`");
});

test("renders block quote with vertical line", () => {
	const input = "> Quote text";
	const root = parse(input, core, false);
	const output = renderToConsole(root);
	const stripped = output.replace(/\x1b\[[0-9;]*m/g, "");
	expect(stripped).toContain("┃ Quote text");
});

test("renders thematic break", () => {
	const input = "---";
	const root = parse(input, core, false);
	const output = renderToConsole(root);
	expect(output).toContain("─");
});

test("renders task list with emojis", () => {
	const input = "- [x] Done\n- [ ] Todo";
	const root = parse(input, gfm, false);
	const output = renderToConsole(root);
	expect(output).toContain("[✓]");
	expect(output).toContain("[ ]");
});

test("renders table with Unicode borders", () => {
	const input = "| A | B |\n|---|---|\n| 1 | 2 |";
	const root = parse(input, gfm, false);
	const output = renderToConsole(root);
	expect(output).toContain("┌");
	expect(output).toContain("┬");
	expect(output).toContain("┐");
	expect(output).toContain("┼");
	expect(output).toContain("│");
	expect(output).toContain("├");
	expect(output).toContain("┤");
	expect(output).toContain("└");
	expect(output).toContain("┴");
	expect(output).toContain("┘");
});

test("renders strong text", () => {
	const input = "**bold**";
	const root = parse(input, core, false);
	const output = renderToConsole(root);
	expect(output).toContain("bold");
});

test("renders emphasis text", () => {
	const input = "*italic*";
	const root = parse(input, core, false);
	const output = renderToConsole(root);
	expect(output).toContain("italic");
});

test("renders link", () => {
	const input = "[text](url)";
	const root = parse(input, core, false);
	const output = renderToConsole(root);
	expect(output).toContain("text");
	expect(output).toContain("url");
});

test("renders image", () => {
	const input = "![alt](url)";
	const root = parse(input, core, false);
	const output = renderToConsole(root);
	expect(output).toContain("Image");
});

test("renders strikethrough", () => {
	const input = "~~deleted~~";
	const root = parse(input, gfm, false);
	const output = renderToConsole(root);
	expect(output).toContain("\x1b[9mdeleted\x1b[29m");
});

test("renders alert with emoji", () => {
	const input = "> [!NOTE]\n> Note content";
	const root = parse(input, gfm, false);
	const output = renderToConsole(root);
	expect(output).toContain("📝");
	expect(output).toContain("Note:");
});

test("renders nested list with different bullets", () => {
	const input = "- Level 1\n  - Level 2\n    - Level 3";
	const root = parse(input, core, false);
	const output = renderToConsole(root);
	expect(output).toContain("• Level 1");
	expect(output).toContain("◦ Level 2");
});
